### PR TITLE
feat(storage): support write anytime and spill to disk

### PR DESCRIPTION
**Pull request by wenym1**
## What's changed and what's your intention?
In this PR, we support write any time in hummock storage. 

All write batches written to shared buffer will be assigned a monotonically increasing `OrderIndex`. A larger order index indicate a later write batch. 

A local flush task will take the oldest continuous write batches as the task payload. For the term continuous, it means no other write batches, ssts or uploading tasks have order index fallen within the min-max order index range of the task payload. The order index of uploading tasks is the smallest order index in the task payload, and the generated SSTs share such order index.

To serve read request, ordered merge iterator is built according to the order index.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
#1842

### CODE REVIEW

File: src/storage/src/hummock/shared_buffer/mod.rs [hunk 1]
```diff
@@ -146,84 +230,161 @@ impl SharedBuffer {
...
+        let keyed_payload = match task_type {
+            UploadTaskType::FlushWriteBatch => {
+                // For flush write batch, currently we only flush the write batches. We first pick
+                // the write batch with the smallest order index, and then start
+                // from this order index, we iterate over all order indexes in
+                // ascending order. We add the write batches to the task payload and
+                // stop when we meet a sst.
+
+                // Keep track of whether the data of an order index is non uploaded local batches
+                let mut order_index_is_non_upload_batch = BTreeMap::new();
```
Comment (hzxa21) [inline src/storage/src/hummock/shared_buffer/mod.rs:267]: Can you add more comments on why we need to store None as the value?

**Review by hzxa21 — COMMENTED** [review]
Rest LGTM.

File: src/storage/src/hummock/shared_buffer/mod.rs [hunk 2]
```diff
@@ -71,29 +79,101 @@ impl UncommittedData {
...
+    uncommitted_data: &OrderSortedUncommittedData,
+    sstable_store: Arc<SstableStore>,
+    stats: Arc<StateStoreMetrics>,
+    local_stats: &mut StoreLocalStatistic,
+    read_options: Arc<ReadOptions>,
+) -> HummockResult<BoxedHummockIterator<T::Direction>> {
+    let mut ordered_iters = Vec::new();
+    // use `rev` because the larger order index comes at first.
+    for data_list in uncommitted_data {
+        let mut data_iters = Vec::new();
```
Comment (BugenZhao) [inline src/storage/src/hummock/shared_buffer/mod.rs:113]: Use `with_capacity` here?

**Review by BugenZhao — COMMENTED** [review]
Good work! Generally LSTM!

File: src/storage/src/hummock/shared_buffer/mod.rs [hunk 3]
```diff
@@ -71,29 +79,101 @@ impl UncommittedData {
...
+
+pub(crate) async fn build_ordered_merge_iter<T: HummockIteratorType>(
+    uncommitted_data: &OrderSortedUncommittedData,
+    sstable_store: Arc<SstableStore>,
+    stats: Arc<StateStoreMetrics>,
+    local_stats: &mut StoreLocalStatistic,
+    read_options: Arc<ReadOptions>,
+) -> HummockResult<BoxedHummockIterator<T::Direction>> {
+    let mut ordered_iters = Vec::new();
+    // use `rev` because the larger order index comes at first.
```
[the remark this answers is not in the document]
Comment (wenym1) [reply src/storage/src/hummock/shared_buffer/mod.rs]: Oh it's just the stale comment for some code I had deleted. I have removed the comment already.

**Pull request merged**